src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -20,83 +15,126 @@ public partial class SocketAsyncEventArgs : EventArgs, IDisposable
         private int _sendPacketsElementsBufferCount;
 
         // Internal variables for SendPackets
         private FileStream[] _sendPacketsFileStreams;
         private SafeHandle[] _sendPacketsFileHandles;
-        private IntPtr _ptrSendPacketsDescriptor;
 
         // Overlapped object related variables.
         private SafeNativeOverlapped _ptrNativeOverlapped;
         private PreAllocatedOverlapped _preAllocatedOverlapped;
-        private object[] _objectsToPin;
+
         private enum PinState
         {
-            None = 0,
-            NoBuffer,
+            NoBuffer = 0,
             SingleAcceptBuffer,
             SingleBuffer,
             MultipleBuffer,
             SendPackets
         }
         private PinState _pinState;
-        private byte[] _pinnedAcceptBuffer;
-        private byte[] _pinnedSingleBuffer;
-        private int _pinnedSingleBufferOffset;
-        private int _pinnedSingleBufferCount;
 
         internal int? SendPacketsDescriptorCount
         {
             get
             {
-                return _sendPacketsDescriptor == null ? null : (int?)_sendPacketsDescriptor.Length;
+                return _sendPacketsElementsInternal == null ? null : (int?)_sendPacketsElementsFileCount + _sendPacketsElementsBufferCount;
             }
         }
 
         private void InitializeInternals()
         {
             // Zero tells TransmitPackets to select a default send size.
             _sendPacketsSendSize = 0;
+
+            AllocateOverlapped();
         }
 
-        private void FreeInternals(bool calledFromFinalizer)
+        private void FreeInternals()
         {
-            // Free native overlapped data.
-            FreeOverlapped(calledFromFinalizer);
+            ClearPinState();
+
+            // Free the overlapped object.
+            if (_ptrNativeOverlapped != null && !_ptrNativeOverlapped.IsInvalid)
+            {
+                _ptrNativeOverlapped.Dispose();
+                _ptrNativeOverlapped = null;
+            }
+
+            // Dispose the pre-allocated overlapped
+            Debug.Assert(_preAllocatedOverlapped != null);
+            _preAllocatedOverlapped.Dispose();
+            _preAllocatedOverlapped = null;
+
+            // Unpin socketAddress, if necessary
+            if (_socketAddressBufferGCHandle.IsAllocated)
+            {
+                Debug.Assert(_socketAddressBufferGCHandle.Target == _socketAddress.Buffer);
+                _socketAddressBufferGCHandle.Free();
+            }
+
+            // Unpin controlBuffer, if necessary
+            if (_controlBufferGCHandle.IsAllocated)
+            {
+                Debug.Assert(_controlBufferGCHandle.Target == _controlBuffer);
+                _controlBufferGCHandle.Free();
+            }
+        }
+
+        private unsafe void AllocateOverlapped()
+        {
+            _preAllocatedOverlapped = new PreAllocatedOverlapped(CompletionPortCallback, this, null);
```

Comment:
This is going to allocate a delegate for CompletionPortCallback on every call to this.